jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java

```diff
@@ -970,38 +971,49 @@ public boolean checkVirtualHost(final Request baseRequest)
         if (_vhosts != null && _vhosts.length > 0)
         {
             String vhost = normalizeHostname(baseRequest.getServerName());
-
-            boolean match = false;
-            boolean connectorName = false;
-            boolean connectorMatch = false;
-
-            for (String contextVhost:_vhosts)
+            String connectorName = baseRequest.getHttpChannel().getConnector().getName();
+            for (String contextVhost : _vhosts)
             {
-                if (contextVhost == null || contextVhost.length()==0)
+                if (contextVhost == null || contextVhost.length() == 0)
                     continue;
-                char c=contextVhost.charAt(0);
-                switch (c)
+
+                int connectorPos = contextVhost.indexOf('@') + 1;
```

Comment:
rather than splitting this on every request, how about splitting it in the setter and then having two matching arrays:
```
    private String[] _vhosts; // Host name portion, matching _vconnectors array
    private String[] _vconnectors; // connector portion, matching _vhosts array
```